### PR TITLE
Obfuscated private ip fix

### PR DESCRIPTION
**Pull request by SparrowBrain**
### Issue
Currently logging into Steam on more than one machine will end up with lost authentication.

Let's assume the scenario:
* We have two PCs sharing a single public IP address (sitting behind a router)
* Log into GOG Galaxy Steam integration plugin on **PC A**
* Log into GOG Galaxy Steam integration plugin on **PC B**
* Restart GOG Galaxy on **PC A**
* Steam authentication now fails on **PC A** for the plugin 

### Implementation
It looks like Steam uses a combination of public IP address and private IP address to uniquely identify the connections. Specifically the login requests have the `obfuscated_private_ip` field. It is widely used by other unofficial Steam integration libraries ([SteamRE](https://github.com/SteamRE/SteamKit/blob/35792093c6ea1475f9c10086d09d9163f5429ef8/SteamKit2/SteamKit2/Steam/Handlers/SteamUser/SteamUser.cs#L327), [ValvePython](https://github.com/ValvePython/steam/blob/44d0ef5cc3f420205e3d893abb362a507dd93af2/steam/client/__init__.py#L542)). Usually the solution is to do IP v4 XOR against a mask.

The obfuscation mask seems to be chosen randomly ([0xF00DBAAD](https://github.com/ValvePython/steam/blob/44d0ef5cc3f420205e3d893abb362a507dd93af2/steam/client/__init__.py#L542) and [0xBAADF00D](https://github.com/SteamRE/SteamKit/blob/62d8d4422cc2e3cfc25bd30e4b194a86a6817daa/SteamKit2/SteamKit2/Base/Generated/SteamLanguageInternal.cs#L660). Hence I took creative liberty to pick `0x606573A4` which is as close I could get to GOG STEAM in leet speak.

The private IP is taken from the currently open socket.

Additionally removed couple of unused methods from `protobuf_client.py`

I have tested the implementation on a local network with 2 computers. Also tested if the authentication keeps working in case the private IP changes. Didn't detect any issues.

### Testing
To test either apply the changes manually, or download [obfuscated_private_ip_fix.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5832230/obfuscated_private_ip_fix.zip) (this is v0.54 with the fix) and replace plugin contents in `%LOCALAPPDATA%\GOG.com\Galaxy\plugins\installed\steam_ca27391f-2675-49b1-92c0-896d43afa4f8`

### CODE REVIEW

**Comment by FriendsOfGalaxy** [comment]
@SparrowBrain nice, thanks! Taking custom mask is brilliant idea, my previous attempt to use MAC address in #81 is a way less safe option.

I'm going to test it asap :)

File: src/protocol/protobuf_client.py [hunk 1]
```diff
@@ -171,6 +151,13 @@ async def log_on_token(self, steam_id, account_name, token, used_server_cell_id)
         logger.info("Sending log on message using token")
         await self._send(EMsg.ClientLogon, message)
 
+    def _get_obfuscated_private_ip(self) -> int:
+        (host, port) = self._socket.local_address
```
Comment (FriendsOfGalaxy) [inline src/protocol/protobuf_client.py:155]:
```suggestion
        host, port = self._socket.local_address
```

**Comment by FriendsOfGalaxy** [comment]
Another test scenario:

- We have two PCs sharing a single public IP address (sitting behind a router)
- Log into GOG Galaxy Steam integration plugin on PC A
- Log into GOG Galaxy Steam integration plugin on PC B
- check logs on PC A - there is eresult 34 coming as notification message from Steam

@SparrowBrain 
> Hence I took creative liberty to pick 0x606573A4 which is as close I could get to GOG STEAM in leet speak.

Could you explain this? Isn't it random mask?

> It looks like Steam uses a combination of public IP address and private IP address to uniquely identify the connections
> The private IP is taken from the currently open socket.

So I in some curcaimstances user will have the two computers in different networks but with the same privet IP, then we should expect the issue to occur?

**Comment by Chupaka** [comment]
> Could you explain this? Isn't it a random mask chosen by you?

[606 = GOG, 573A4 = STEAM](https://en.wikipedia.org/wiki/Leet) :)

**Comment by FriendsOfGalaxy** [comment]
Ah, I see:)

**Comment by Chupaka** [comment]
> So if in some curcaimstances user will have the two computers in different networks but with the same privet IP, then we should expect the issue to occur?

Good point. I'd recommend some hash of mac+ip then. And that magic constant, of course :)

**Comment by FriendsOfGalaxy** [comment]
> UPD: but... Won't they use different logins? :)

yes, but the issue is that steam backend invalidates one session when another is created from similar LogOn message. We don't exact logic though.

In fact - we have a clear reproduction only in one and the same network.

---

I tested this build by loging on 2 machines multiple time and it looks quite solid.

Once I got an error that NoneType has no local_address, but that would failed anyway, as the code tries to proceed with login without websocket connection established (looks like race condition when connection is reseting). But after restarting Galaxy it has been awaken correctly.

**Review by FriendsOfGalaxy — APPROVED** [review]

File: src/protocol/protobuf_client.py (hunk 1 above)
Comment (SparrowBrain) [reply src/protocol/protobuf_client.py]: Good catch.

**Comment by SparrowBrain** [comment]
> Another test scenario:
> 
>     * We have two PCs sharing a single public IP address (sitting behind a router)
> 
>     * Log into GOG Galaxy Steam integration plugin on PC A
> 
>     * Log into GOG Galaxy Steam integration plugin on PC B
> 
>     * check logs on PC A - there is eresult 34 coming as notification message from Steam

Is this still something we should worry about? Sorry, I didn't have time to do any tests recently.

If it is, we could potentially create a separate issue to resolve it, since it looks like `obfuscated_private_ip` is still useful and could help people keep their authentication.

**Comment by ReiTung** [comment]
I think the above (multiple PCs behind one priv IP) is a rather common scenario. I suffered it on my own and also pointed that out as part of some testing for #74 ...
I finally "mitigated" by disabling Steam Integration on all clients beside one, as GOG seems to synchronize it's database across the machines anyway (via the GOG servers though, I do not think it has a local peer functionality built in). Eventually, I will get the data to all machines once synced on the one.
Might be worth fixing (and I agree a separate issue to better track makes sense), but personally, I would give it a Prio 2 compared to the major issues and also given there is a workaround.

Just my few cents

**Comment by FriendsOfGalaxy** [comment]
>I think the above (multiple PCs behind one priv IP) is a rather common scenario.

@ReiTung you probably mean public IP. Your router gives different privet IPs for each your machine.

**Comment by FriendsOfGalaxy** [comment]
@SparrowBrain that is exactly the repro your change should fix. From what I've found the one you've mention on PR description is a side effect of that situation (so should be fixed as well). 

If you or anybody could test something else I would suggest case where you have 2 computers on 2 networks with the same private IP. Just to make sure that Steam does not switch to distinguish sessions ONLY by checking `obfuscated_privat_ip` if given. I think it should take public IP into account, but we don't know for sure.

I've merged, but I'm going to release it early next week.

**Comment by ReiTung** [comment]
@FriendsOfGalaxy my bad, of course, typo. External IP is public, internals private ... A scenario I am definitely running and able to test if needed

**Comment by FriendsOfGalaxy** [comment]
@ReiTung great! Please test the build from description using test scenario from test description as well. Make sure private IPs on both machines are the same (like 192.168.1.1)

**Comment by FriendsOfGalaxy** [comment]
released

**Comment by ReiTung** [comment]
> @ReiTung great! Please test the build from description using test scenario from test description as well. Make sure private IPs on both machines are the same (like 192.168.1.1)

do you want to have both testcases, i.e. different private IPs on the two machines as well as the same private IP for both? or just the latter one?

**Comment by FriendsOfGalaxy** [comment]
@ReiTung just the latter one

**Comment by ReiTung** [comment]
Hi Team,
so, did some testing - seems using this fix does not create the old error message on the primary PC anymore when using the same IP on the second PC. However, I am still receiving a lot of invalid logons on the primary PC (error code 5), but cannot link them to the second PC being online. Gut feeling is that logon issues (error 5) increased in total with v0.55 special edition (#108+server cache+caching steam content), but these also seem to occur with second PC totally off - so not sure if linked.

logs of PC B attached (can attach PC A as well if needed)

thanks Rei

[logs_PC_B_v055_special_edition_20210211.zip](https://github.com/FriendsOfGalaxy/galaxy-integration-steam/files/5998325/logs_PC_B_v055_special_edition_20210211.zip)

**Pull request merged**